gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/ProjectInfo.java

```diff
@@ -0,0 +1,238 @@
+      return this;
+    }
+
+    public Builder removeLabel(String key) {
+      this.labels.remove(key);
+      return this;
+    }
+
+    public Builder clearLabels() {
+      this.labels.clear();
+      return this;
+    }
+
+    public Builder labels(Map<String, String> labels) {
+      this.labels = checkNotNull(labels);
+      return this;
+    }
+
+    Builder number(Long number) {
+      this.number = number;
+      return this;
+    }
+
+    Builder state(State state) {
+      this.state = state;
+      return this;
+    }
+
+    Builder createTimeMillis(Long createTimeMillis) {
+      this.createTimeMillis = createTimeMillis;
+      return this;
+    }
+
+    public Builder parent(ResourceId parent) {
+      this.parent = parent;
+      return this;
+    }
+
+    public ProjectInfo build() {
+      return new ProjectInfo(name, id, labels, number, state, createTimeMillis, parent);
+    }
+  }
+
+  ProjectInfo(String name, String id, Map<String, String> labels, Long number, State state,
+      Long createTimeMillis, ResourceId parent) {
+    this.name = name;
+    this.id = checkNotNull(id);
+    ImmutableMap.Builder<String, String> labelsMapBuilder = ImmutableMap.builder();
+    for (Map.Entry<String, String> entry : labels.entrySet()) {
+      labelsMapBuilder.put(entry.getKey(), entry.getValue());
+    }
+    this.labels = (labels == null) ? null : labelsMapBuilder.build();
+    this.number = number;
+    this.state = state;
+    this.createTimeMillis = createTimeMillis;
+    this.parent = parent;
+  }
+
+  public String id() {
+    return id;
+  }
+
+  public String name() {
+    return name;
+  }
+
+  public Long number() {
+    return number;
+  }
+
+  public Map<String, String> labels() {
+    return labels;
+  }
+
+  public State state() {
+    return state;
+  }
+
+  public Long createTimeMillis() {
+    return createTimeMillis;
+  }
+
+  public ResourceId parent() {
+    return parent;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj == this) {
+      return true;
+    } else if (obj instanceof ProjectInfo) {
+      ProjectInfo other = (ProjectInfo) obj;
+      return Objects.equals(this.name, other.name) && Objects.equals(this.id, other.id)
+          && Objects.equals(this.labels, other.labels) && Objects.equals(this.number, other.number)
+          && Objects.equals(this.state, other.state)
+          && Objects.equals(this.createTimeMillis, other.createTimeMillis)
+          && Objects.equals(this.parent, other.parent);
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(name, id, labels, number, state, createTimeMillis, parent);
+  }
+
+  public static Builder builder(String id) {
+    return new Builder().id(id);
+  }
+
+  public Builder toBuilder() {
+    Map<String, String> mutableLabels = new HashMap<String, String>();
+    for (Map.Entry<String, String> entry : labels.entrySet()) {
+      mutableLabels.put(entry.getKey(), entry.getValue());
+    }
```

Comment:
If we take my suggestion of creating a new `HashMap` inside `builder.labels` setter then `mutableLabels` here is no longer necessary.